File: Go/src/0101 ~ 0200/0119 - Pascal's Triangle II.go
// 链接：https://leetcode.com/problems/pascals-triangle-ii/
// 题意：给定数字 k ，返回杨辉三角的第 k 层（使用 O(k) 的空间）？

// 输入： 3
// 输出： [1,3,3,1]

// 思路：DP
//
//		思路和 0118 一致，
//		由于只需要最后一层的数据，所以我们完全可以用一维数组，
//		利用原有的数据直接更新即可
//
//		时间复杂度： O(k ^ 2) 【当然可以利用组合数公式直接计算，进一步降低复杂度】
//		空间复杂度： O(k)

func getRow(k int) []int {
	result := make([]int, k + 1)
	result[0] = 1
	// 处理后面的行
	for i := 1; i <= k; i++ {
		// 最后一个数字为 1
		result[i] = 1

		// 通过上一行的数字更新本行的数字
		// 由于后一个需要知道前一个的上一行数字，所以从后往前更新
		for j := i - 1; j > 0; j-- {
			result[j] += result[j - 1]
		}
	}
	return result
}

File: Rust/src/0101 ~ 0200/0119 - Pascal's Triangle II.rs
// 链接：https://leetcode.com/problems/pascals-triangle-ii/
// 题意：给定一个整数 rowIndex ，返回杨辉三角的第 rowIndex 行。
//
//      在杨辉三角中，每一个数是它左上方和右上方的数之和。
//
//      进阶：使用空间复杂度为 O(n) 的方法。


// 数据限制：
//  1 <= rowIndex <= 33


// 输入： rowIndex = 3
// 输出： [1,3,3,1]

// 输入： rowIndex = 0
// 输出： [1]

// 输入： rowIndex = 1
// 输出： [1,1]


// 思路： 模拟
//
//      本题是 LeetCode 118 这题的加强版，需要将空间复杂度优化为 O(n) 。
//
//      空间复杂度为 O(n ^ 2) 的方法很简单，按照题意从第一层开始计算即可，
//      对于每一个位置 (i, j) 有 dp[i][j] = dp[i - 1][j - 1] + dp[i][j] 。
//
//      注意边界情况，当 j == 0 || j == i 时， dp[i][j] = 1 。
//
//      DP 常见的三种优化方式见 LeetCode 583 这题的思路，
//      本题直接使用一维数组 + 倒序转移这种方法进行优化。 
//
//      因为状态 dp[i][j] 仅由 dp[i - 1][..=j] 中的状态转移而来时，
//      可以使用倒序转移，从后往前计算，以免使用当前行的状态进行转移。
//
//
//      时间复杂度：O(n ^ 2)
//          1. 需要遍历计算全部 O(n ^ 2) 个位置的数
//      空间复杂度：O(n)
//          1. 需要用 dp 维护全部 O(n) 个位置的数


impl Solution {
    pub fn get_row(row_index: i32) -> Vec<i32> {
        // 默认全都是 1 ，直接所有行的边界情况设置好
        let mut dp = vec![1; row_index as usize + 1];
        for i in 0..=row_index as usize {
            // 倒序转移，仅处理第 i 层中间非边界情况的数，即它等于左上和右上的数之和
            for j in (1..i).rev() {
                dp[j] += dp[j - 1];
            }
        }

        dp
    }
}
